File: internal/domain/repositories.go
package domain

import "context"

type UserRepository interface {
	GetUsers(ctx context.Context) ([]User, error)
	GetUser(ctx context.Context, id string) (User, error)
	GetByUsername(ctx context.Context, username string) (User, error)
	CreateUser(ctx context.Context, user User) error
	UpdateUser(ctx context.Context, user User) error
	UpdateUserStatus(ctx context.Context, id string) error
	DeleteUser(ctx context.Context, id string) error
}

type PrizeRepository interface {
	GetPrizes(ctx context.Context) ([]Prize, error)
	GetPrize(ctx context.Context, id string) (Prize, error)
	GetPrizesByYear(ctx context.Context, year string) ([]Prize, error)
	GetPrizesByCategory(ctx context.Context, category string) ([]Prize, error)
	GetPrizesByCategoryAndYear(ctx context.Context, category string, year string) ([]Prize, error)
	GetCategories(ctx context.Context) ([]string, error)
	GetYears(ctx context.Context) ([]string, error)
}

File: src/domain/repositories.rs
use async_trait::async_trait;
use crate::domain::models::{User, Prize};

#[async_trait]
pub trait UserRepository: Send + Sync {
    async fn get_users(&self) -> Result<Vec<User>, Box<dyn std::error::Error + Send + Sync>>;
    async fn get_user(&self, id: &str) -> Result<User, Box<dyn std::error::Error + Send + Sync>>;
    async fn get_by_username(&self, username: &str) -> Result<User, Box<dyn std::error::Error + Send + Sync>>;
    async fn create_user(&self, user: User) -> Result<(), Box<dyn std::error::Error + Send + Sync>>;
    async fn update_user(&self, user: User) -> Result<(), Box<dyn std::error::Error + Send + Sync>>;
}

#[async_trait]
pub trait PrizeRepository: Send + Sync {
    async fn get_prizes(&self) -> Result<Vec<Prize>, Box<dyn std::error::Error + Send + Sync>>;
    async fn get_prize(&self, id: &str) -> Result<Prize, Box<dyn std::error::Error + Send + Sync>>;
    async fn get_prizes_by_year(&self, year: &str) -> Result<Vec<Prize>, Box<dyn std::error::Error + Send + Sync>>;
    async fn get_prizes_by_category(&self, category: &str) -> Result<Vec<Prize>, Box<dyn std::error::Error + Send + Sync>>;
    async fn get_prizes_by_category_and_year(&self, category: &str, year: &str) -> Result<Vec<Prize>, Box<dyn std::error::Error + Send + Sync>>;
    async fn get_categories(&self) -> Result<Vec<String>, Box<dyn std::error::Error + Send + Sync>>;
    async fn get_years(&self) -> Result<Vec<String>, Box<dyn std::error::Error + Send + Sync>>;
}
